### PR TITLE
Do not initialize FS-watching in CLI and TAPI

### LINKED ISSUE #17128: Do not load file-events library in the tooling API client
Currently, we load the file events library on the JVM which uses the TAPI client and in the command line client. For the command line client, that doesn't make a big difference, though for the TAPI client it causes the JVM which uses it to keep a reference to a classloader reference, since we store references to the classes in native code when loading the JNI library.

This may cause memory problems in the TAPI client, as seen here: https://github.com/gradle/gradle-private/issues/3355

The client doesn't need the file events library, so it shouldn't load it.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by wolfs**
Do not initialize file system watching (aka loading the fs-events library) in the client JVMs for the CLI and tooling API launchers. Note that this disables file system watching when running with `--no-daemon`, since then the CLI jvm is re-used.

Fixes #17128

### CODE REVIEW

**Comment by wolfs** [comment]
This seems to work.

@lptr This clashes somewhat with your PR https://github.com/gradle/gradle/issues/17376. Now we would have another reason why we don't watch the file system, even though watching has been requested: When we don't use a daemon JVM. We need to decide what to log in this case.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java [hunk 1]
```diff
@@ -120,22 +169,17 @@ public static synchronized void initialize(File userHomeDir, boolean initializeA
...
-                        }
-                        JANSI_BOOT_PATH_CONFIGURER.configure(nativeBaseDir);
-                    }
                     LOGGER.info("Initialized native services in: {}", nativeBaseDir);
                 }
                 initialized = true;
             }
+            for (NativeFeatures requestedFeature : requestedFeatures) {
+                if (!INITIALIZED_FEATURES.contains(requestedFeature)) {
+                    INITIALIZED_FEATURES.add(requestedFeature);
```
Comment (lptr) [inline subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java:178]: I find this weird with capitals like this. Maybe we have too much static state here, and should have a static instance of a new thing that has instance fields instead?

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java [hunk 2]
```diff
@@ -80,22 +81,70 @@
...
     private static boolean initialized;
+    private static final EnumSet<NativeFeatures> INITIALIZED_FEATURES = EnumSet.noneOf(NativeFeatures.class);
 
     public static final String NATIVE_DIR_OVERRIDE = "org.gradle.native.dir";
 
+    public enum NativeFeatures {
+        FILE_SYSTEM_WATCHING {
+            @Override
+            public void initialize(File nativeBaseDir) {
+                if (useNativeIntegrations) {
```
Comment (lptr) [inline subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java:92]: Can this `if` be moved to where we create the enum set instead?

I find it hard to follow the value of `useNativeIntegrations`, maybe if it wasn't overwritten it would be easier to follow the logic?

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerFileSystemWatchingIntegrationTest.groovy [hunk 3]
```diff
@@ -38,6 +37,7 @@ class GradleRunnerFileSystemWatchingIntegrationTest extends BaseGradleRunnerInte
         assumeTrue("File system watching is enabled by default", gradleVersion >= GradleVersion.version("7.0"))
     }
 
+    @NoDebug
```
Comment (lptr) [inline subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerFileSystemWatchingIntegrationTest.groovy:40]: Why is this needed?

**Review by lptr — COMMENTED** [review]
I like the direction, left a few comments.

File: subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerFileSystemWatchingIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/test-kit/src/integTest/groovy/org/gradle/testkit/runner/GradleRunnerFileSystemWatchingIntegrationTest.groovy:40]: This test only works when using `debug = false` on `GradleRunner`. I moved the `@NoDebug` from the class to the individual methods so I can also have a test for the behavior with `debug = true` in the same class.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java (hunk 1 above)
Comment (wolfs) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java]: I moved most of the static fields to the actual `NativeServices` instance. I think this makes the code easier to understand.

File: subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java (hunk 2 above)
Comment (wolfs) [reply subprojects/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java:93]: I now pass in the `useNativeIntegrations` into this method. I think it is easier to understand now.

**Comment by wolfs** [comment]
@lptr PTAL!

**Review by lptr — APPROVED** [review]
LGTM.

What is stopping us from making `NativeServices` a real service, initialized and injected the normal way instead of this static initializaiton?

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Comment by wolfs** [comment]
> What is stopping us from making NativeServices a real service, initialized and injected the normal way instead of this static initializaiton?

There are a few locations where we call `NativeServices.getInstance()`, I don't know exactly why we can't use injection there.
